File: src/engine/orchestrator.ts
import type { JobDefinition } from "../domain/job-definition.js";
import type { WorkflowInstance } from "../domain/workflow-instance.js";
import { WorkflowState } from "../domain/types.js";
import type { DomainEventListener } from "../domain/events.js";
import type { WorkflowStore, RunStore } from "../storage/interfaces.js";
import { WorkflowStateMachine } from "./state-machine.js";
import { TransitionGuard } from "./transition-guard.js";
import type { Clock } from "../utils/clock.js";
import type { Logger } from "../utils/logger.js";

export class WorkflowOrchestrator {
  private listeners: DomainEventListener[] = [];

  constructor(
    private readonly stateMachine: WorkflowStateMachine,
    private readonly guard: TransitionGuard,
    private readonly workflowStore: WorkflowStore,
    private readonly runStore: RunStore,
    private readonly clock: Clock,
    private readonly log: Logger,
  ) {}

  onEvent(listener: DomainEventListener): void {
    this.listeners.push(listener);
  }

  private emit(event: Parameters<DomainEventListener>[0]): void {
    for (const l of this.listeners) {
      l(event);
    }
  }

  async applyTransition(
    instance: WorkflowInstance,
    event: string,
    target: WorkflowState,
    actor: string,
  ): Promise<WorkflowInstance> {
    this.guard.evaluate({ instance, event, target });
    const next = this.stateMachine.transition(instance, event, target, actor);
    await this.workflowStore.save(next);

    const run = await this.runStore.get(next.runId);
    if (run) {
      run.status = target;
      if (this.stateMachine.isTerminal(target)) {
        run.finishedAt = this.clock.nowIso();
      }
      await this.runStore.save(run);
    }

    this.emit({
      type: "workflow.transitioned",
      runId: next.runId,
      from: instance.state,
      to: target,
      event,
      at: this.clock.nowIso(),
    });

    this.log.info("workflow transitioned", {
      workflowId: next.id,
      from: instance.state,
      to: target,
      event,
    });

    return next;
  }

  async startRun(
    _definition: JobDefinition,
    workflow: WorkflowInstance,
  ): Promise<WorkflowInstance> {
    let current = workflow;
    if (current.state === WorkflowState.Pending) {
      current = await this.applyTransition(current, "start", WorkflowState.Running, "orchestrator");
    }
    return current;
  }
}

File: .fieldops_engine_snapshot/fieldops/engine/orchestrator.py
from fieldops.domain.enums import WorkflowState, TaskStatus
from fieldops.domain.models import WorkflowInstance
from fieldops.engine.state_machine import WorkflowStateMachine
from fieldops.storage.memory import MemoryWorkflowStore
from fieldops.utils.clock import SystemClock

class WorkflowOrchestrator:
    def __init__(self, store: MemoryWorkflowStore, clock: SystemClock | None = None) -> None:
        self._store = store
        self._sm = WorkflowStateMachine()
        self._clock = clock or SystemClock()

    async def apply(self, wf: WorkflowInstance, event: str) -> WorkflowInstance:
        if not self._sm.can_transition(wf.state, event):
            raise ValueError(f"cannot apply {event} from {wf.state}")
        wf.state = self._sm.transition(wf.state, event)
        wf.updated_at = self._clock.now_iso()
        await self._store.save(wf)
        return wf

    async def mark_task(self, wf: WorkflowInstance, task_id: str, status: TaskStatus, output: dict | None = None, error: str | None = None) -> WorkflowInstance:
        for t in wf.tasks:
            if t.definition_task_id == task_id:
                t.status = status
                t.finished_at = self._clock.now_iso() if status in (TaskStatus.SUCCEEDED, TaskStatus.FAILED) else t.finished_at
                if output:
                    t.output = output
                if error:
                    t.error_message = error
        wf.updated_at = self._clock.now_iso()
        await self._store.save(wf)
        return wf
